File: leetcode/59. Spiral Matrix II.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        int max_x = n,
            max_y = n,
            min_x = 0,
            min_y = 0,
            num = 1;
        vector<vector<int> > ans(n, vector<int>(n));
        while (min_x < max_x and min_y < max_y) {
            // right
            for (int i = min_y; i < max_y; i++) {
                ans[min_x][i] = num;
                num++;
            }
            min_x++;
            // down
            for (int i = min_x; i < max_x; i++) {
                ans[i][max_y - 1] = num;
                num++;
            }
            max_y--;
            // left
            // check if there is a row left
            if (min_x < max_x) {
                for (int i = max_y - 1; i >= min_y; i--) {
                    ans[max_x - 1][i] = num;
                    num++;
                }
                max_x--;
            }
            // up
            // check if there is a column left
            if (min_y < max_y) {
                for (int i = max_x - 1; i >= min_x; i--) {
                    ans[i][min_y] = num;
                    num++;
                }
                min_y++;
            }
        }
        return ans;
    }
};

File: leetcode/59. Spiral Matrix II.js
/**
 * @param {number} n
 * @return {number[][]}
 */
var generateMatrix = function(n) {
    var max_x = n,
        max_y = n,
        min_x = 0,
        min_y = 0,
        num = 1;
    var ans = new Array(n);
    for (let i = 0; i < n; i++) {
        ans[i] = new Array(n);
    }
    while (min_x < max_x && min_y < max_y){
        for (let i = min_y; i < max_y; i++) {
            ans[min_x][i] = num;
            num++;   
        }
        min_x++;
        for (let i = min_x; i < max_x; i++) {
            ans[i][max_y - 1] = num;
            num++;
        }
        max_y--;
        if (min_x < max_x){
            for (let i = max_y - 1; i >= min_y; i--) {
                ans[max_x - 1][i] = num;
                num++;
            }
            max_x--;
        }
        if (min_y < max_y){
            for (let i = max_x - 1; i >= min_x; i--) {
                ans[i][min_y] = num;
                num++;
            }
            min_y++;
        }
    }
    return ans;

};